README
	geo-*		Rick Richardson's Linux tools for geocaching

Here are various tools for geocaching.  Some are designed to turn the
clickly‐clicky‐scroll‐clicky‐clicky messes that are geocaching.com,
opencaching.com, www.opencaching.us and navicache.com into a set of tools
which you can use in a batch or cron mode to establish your normal caching
routine, backup your caches and cache logs, etc.  Others are used to enter
new waypoints, free geocoder, create custom maps of waypoints, etc.

These tools work on Linux and CygWin on MS Windows and might also work on
Mac OS X.  I have no plans to port them to native Windows, and to
protect your karma I urge others to refrain from doing so.

All of the tools that were written by me are released into the public
domain, with no copyright or license restrictions whatsoever.  Other
tools, like gpx2html and geodetics.html, are copyright by their
original author and may have license restrictions as indicated in the
source code.

Any of the tools will print a usage summary with "-?", e.g.
    $ geo-gid -?

geo-*.tar.gz
	Tarball of everything below and more.
	    To install:
	    1)	$ make

	    2)	$ make install			# to install in $HOME/bin
		OR
		# PREFIX=/usr make install	# to install in /usr/bin

	    3) Create a $HOME/.georc file with at least these lines in it:

		PASSWORD="your-gc.com-password"
		USERNAME="your-gc.com-username"
		LAT=44.497250	# Your home latitude in decimal degrees
		LON=-93.941033	# Your home longitude in decimal degrees

Tools for accessing gc.com...
	(PMO) : this program works for gc.com "Premium Member Only" subscribers

	geo-found	List caches found (by you or someone else)
	geo-nearest	List the nearest caches to a location
	geo-newest	List the newest caches in a state
	geo-placed	List caches placed (by you or someone else)
	geo-keyword	List caches by keywords.
			All of the above can enter the waypoints into the
			GpsDrive MySQL database.
	geo-html2gpx	Convert a gc.com printable web page (such as the
			above commands can produce with the -H option) to
			a GPX file.

	geo-count	Count caches found
	geo-usernum	Determine gc.com user number (used by geo-count)

(PMO)	geo-gid		Retrieve cache info by GCxxxx waypoint name
(PMO)	geo-gpx		Retrieve GPX file by GCxxxx waypoint name
(PMO)	geo-demand	Request an immediate pocket query email
(PMO)	geo-gpxmail	Process PQ email using gpx2html
(PMO)	geo-gpxprocess	Process PQ download(s) using geo-pqdownload and gpx2html
(PMO)	geo-pqdownload	Perform a Pocket Query download(s)
(PMO)	geo-myfinds	Schedule a Pocket Query containing your finds.
(PMO)	geo-rehides	From your found.gpx file, produce a GPX file of rehides
(PMO)	geo-correct-coords Correct the coords of cache(s)
	geo-density	Compute cache density of an area

	gpx2html	Lightly hacked converter from GPX to HTML
			Originally by fizzymagic (v1.90).  My version
			fixes issues with HTML in the cache descriptions, adds
			sort by latest log date for easy perusing of recent
		        cache activity, and fixes bug in GC[1-9]xxxx.
	gpx-loghistory	Print all logs in reverse cron order.
	geo-pqs		Figure out what PQs to run to get an entire state.
	geo-state	Convenience script;  geo-state -? gives usage.
	geo-sdt         Replace Size, Difficulty, Terrain from a PQ file
	geo-suffix	Replace name with name/TypeSizeDiffTerr/gcid/LatLon
	geo-uniq	Unique the tabsep database
			 
Tools for accessing opencaching.com...
	oc-nearest	List the nearest caches to a location
	oc-newest	List the newest caches in a state
			EXPERIMENTAL, subject to drastic changes

Tools for accessing opencaching.us (and .nl, .de,...) ...
	ok-nearest	List the nearest caches to a location
	ok-newest	List the newest caches in a state
			EXPERIMENTAL, subject to drastic changes

Tools for accessing navicache.com...
	nc-nearest	List the nearest caches to a location
	nc-newest	List the newest caches in a state
			EXPERIMENTAL, subject to drastic changes

Tools for general use
	geo-2gpsdrive	Enter a waypoint file into the GpsDrive MySQL database
	geo-2tangogps	Enter a waypoint file into the tangoGPS or FoxtrotGPS
			sqlite database
	geo-circles	Compute the intersection of two circles on the earth
	geo-trilateration Compute the intersection of three circles on the earth
	geo-triangulation Compute the point by two or three bearings
	geo-intersect	Compute the intersection of two line segments
	geo-polygon	Compute the centroid of a polygon
	geo-gccode2id   Convert GC codes to the decimal equivalent
	geo-id2gccode	Convert decimal IDs to GC codes
	geo-loran-c     Brute force solve of Loran-C problems
	geo-project	Project a waypoint
	geo-code	Geocode an address
	geo-dist	Compute distance along a list of waypoints.
	geo-waypoint	Enter a waypoint into the GpsDrive MySQL database
	geo-map		Create a map with waypoints plotted on it
			These CANNOT be used for publication unless the
			selected map source is the tiger, topographic, or
			aerial map server!
	geo-firefox     Display a map of a point using MapQuest aerial photos
	gpx-finders	Output the finders from a GPX file.
	gpx-merge       GPX file merge.
	gpx-photos	Fetch hi-res aerial photos of all caches in a GPX file
	gpx-stats	Compute stats from a GPX file.
	gpx-unfound     Filter a GPX file removing found caches.

	geodetics.html	A modified version of Gary Nicholson's javascript
			Geodetics Calculator.

Tools for coordinate conversions
	geo-addsub	Add or subtract a value from the coordinates
	geo-coords	Convert lat/lon from one format to another
	geo-incomplete-coords	Print out incomplete coordinates
	ll2maidenhead	Lat/long to Maidenhead (Grid Squares)
	maidenhead2ll	Maidenhead (Grid Squares) to Lat/long
	ll2ggl/ggl2ll	To/From lat/lon to Google Maps QRST string
	ll2osg/osg2ll	To/From lat/lon to British National Grid
	ll2rd/rd2ll	To/From lat/lon to RD (Dutch)
	ll2usng/usng2ll	To/From lat/lon to US National Grid
	ll2utm/utm2ll	To/From lat/lon to UTM
	bing2ll		Bing maps quadkey string to lat/lon
	ll2geohash/geohash2ll	To/From lat/lon to geohash

Tools for manipulating Mapopolis place guide data
	geo-poi		Search place guide (*.pdb or *.csv) for places
	pgpdb2txt	Convert a place guide to plain text

Tools for use by the MN Geocaching Association
	mngca		Count caches found/placed by MnGCA members
	mngca-logs	Create web pages of recent area logs from GPX files
	mngca-newmap	Create newest cache maps for Minnesota

Miscellaneous
	add-pyramids	Add the Pyramids
	adddigits	Add individual digits in a number
	addletters	Add all letters: a=1, b=2, c=3, ... z=26
	anybase2anybase	base to base conversion to/from base 2 thru 62
	atomic-symbol-to-atomic-number Atomic Symbol to Atomic Number
	atomic-symbol-to-period-or-group Atomic Symbol to Period or Group
	atomic-number-to-text Atomic Number to text
	baconian2text	Convert baconian to text
	geo-bacon	Baconian decoder from HTML font's or b's
	balanced-ternary Convert balanced-ternary to/from decimal
	decimal2cryptogram Anthing to cryptogram
	smilies2cryptogram Geocaching 'smilies' to cryptogram
	lethist		Compute letter histogram.
	braille2text	Braille to text translator
	fibonacci-coding Convert a binary coding to a number
	geo-algebraic-expressions Solve a system of algebraic expressions
	geo-alphametic	Solve a math puzzle in which letters stand for digits
	geo-battleship	Map the geocheck.org battleship locations
	geo-clock-angle	Compute the clock angle or the time
	geo-compare-images Compare two images
	geo-char-at	Pick the char at position "n"
	geo-fax		Decode a FAX using 0s and 1s
	geo-lewis-and-clark Encode/decode Lewis and Clark cipher
	geo-math-functions Do various math functions
	geo-morse	Morse decoder
	geo-phone2word	Convert numerical "phone" to a word(s)
	geo-rotate-text	Rotate text CW, CCW, or flip
	geo-slash-backslash Decrypt slashes and backslashes a.k.a Tomtom code
	geo-slash-pipe	Decrypt slash and pipe code
	geo-sub		Do a substitution (caesar) cipher for all shifts
	geo-excel2qrcode Excel to binary or QR barcode converter
	geo-text2qrcode	ASCII text to QR barcode converter
	geo-text2numbers grep for numbers in text
	geo-thumbnails	Recursively extract image thumbnails
	geo-timed-cache Timed cache password fetcher
	geo-wordsearch  Perform a Word Search
	geo-zipcode	Translate zip code to city and state
        mayan-long-count Mayan long count
        navaho-code-talkers Translate Navaho into English
	negadecimal	Convert to/from negabinary, negadecimal
	radio-orphan-annie Radio Orphan Annie’s Decoder Ring
	reverse-montage	reverse (split up) montage image
	segment2text	N-Segment Display to text
	spiritdvd2text	Spirit DVD Code to/from text
	stickman2text	Stickman to text
	tap-code	Tap Code or Polybius Square decoder

Wherigo
	reverse-wherigo Reverse Wherigo decoder
	urwigo-decode	urwigo and earwigo decoder
	wherigo2jpg	Pull jpg images out of a wherigo file
	wherigo2lua	De-compile wherigo
	zonepoint2map	Convert Wherigo ZonePoint's to geo-map coords

Nonograms
	Need to get nonogram solvers from an offsite place!
	geo-nonogram	Nonogram solver
	nono2teal	Convert .nono to teal format
	nono2cross+a	Convert .nono to monochrome cross+a format
	nono2jsolver	Convert .nono to jsolver format
	pbnsolve-wrapper Wrapper for .non format nonograms

GPSBABEL
--------
    Use the latest gpsbabel (1.5.4+):

    $ git clone https://github.com/gpsbabel/gpsbabel.git
    $ cd gpsbabel
    $ ./configure --prefix=/usr
    $ make
    # make install

MAC OSX SUPPORT
---------------

Do this FIRST:
Download and install MacPorts:
    http://www.macports.org/

    $ PATH=$PATH:/opt/local/bin
    $ sudo port selfupdate
    $ sudo port install gsed
    $ sudo port install coreutils
    $ sudo port install lynx
    $ sudo port install ImageMagick
    $ sudo port install gawk
    $ sudo port install ghostscript
    $ sudo port install dos2unix
    $ sudo port install p5-xml-twig
    $ sudo port install p5-datetime
    $ sudo port install p5-html-parser

Download and install gpsbabel:
    http://www.gpsbabel.org/

    $ wget http://www.linklevel.net/distfiles/gpsbabel-1.5.4.tar.gz
    $ tar zxf gpsbabel-1.5.4.tar.gz
    $ cd gpsbabel-1.5.4
    $ ./configure
    $ make
    $ sudo make install

If you get "configure: error: *** A compiler with support for C++11
language features is required.", then drop back and use:

    $ wget http://www.linklevel.net/distfiles/gpsbabel-1.4.4.tar.gz

CYGWIN SUPPORT
--------------

Install CygWin as per: http://www.cygwin.com.  For the easiest (but
longer) install experience, just install everything.  If you choose
to install a subset, you will need to install the base package plus
these other packages (this list might be incomplete):

    ImageMagick     bash                bc              binutils
    cpio            curl                curl-devel      dos2unix
    gawk            gcc                 ghostscript     grep
    libQt5Core      libQt5Core-devel    lynx            make
    sh-utils        sharutils           wget            expat
    units

Download and install gpsbabel:
    $ wget http://www.linklevel.net/distfiles/gpsbabel-1.5.4.tar.gz
    $ tar zxf gpsbabel-*.tar.gz
    $ cd gpsbabel-1.5.4
    $ ./configure --without-libusb
    $ make
    $ make install

FREEBSD SUPPORT
---------------

Install with "pkg install":
    coreutils (gives gtouch and gdate, and more) 
    gawk
    gmake
    gpsbabel
    gsed
    unix2dos

and:
    $ su
    # ln -s /usr/local/bin/bash /bin/bash

Then follow the Linux instructions but use "gmake" instead of "make".


ChangeLog
2018-11-09      Rick Richardson <rick.richardson@comcast.net>
	* anybase2anybase: Add -k and -K flags

2018-11-09      Rick Richardson <rick.richardson@comcast.net>
	* geo-gpx: GC.com changed protocol. Change: 
		    -d ctl00\$ContentBody\$btnGPXDL="GPX+File" \
		To:
		    -d __EVENTTARGET=ctl00\$ContentBody\$lnkGpxDownload \

2018-10-13      Rick Richardson <rick.richardson@comcast.net>
	* geo-common-gc: GC.com changed protocol.
		In gc-login, change:
			Username
		To:
			UsernameOrEmail

2018-09-29      Rick Richardson <rick.richardson@comcast.net>
	* geo-map: Change the default to -a leaflet
		This is because Google Maps recently puts:
		    "For development purposes only"
		on a map if you don't have an API Key.
	* geo-map: Add GMAP_KEY to the ~/.georc file in case you want to pay
		for google maps. Use MAPSRC=gmap in .georc.

2018-09-16      Dylan Richardson <dylanjackrichardson@gmail.com>
	* geo-map: Added Leaflet code, similar to -agmap.
		geo-map -a50 OR geo-map -aleaflet

2018-08-29      Rick Richardson <rick.richardson@comcast.net>
	* New program: radio-orphan-annie: Radio Orphan Annie’s Decoder Ring

2018-08-14      Rick Richardson <rick.richardson@comcast.net>
	* geo-common-gc:	GC.com changed protocol.
		Change:
		        URL="$GEOS/account/login"
		To:
			URL="$GEOS/account/signin"

2018-04-12      Rick Richardson <rick.richardson@comcast.net>
2018-10-23      Rick Richardson <rick.richardson@comcast.net>
	* geo-map: Change -a gstatic to use an API key=xxx

2018-10-13      Rick Richardson <rick.richardson@comcast.net>
	* geo-common-gc: GC.com changed protocol.
		In gc-login, change:
			Username
		To:
			UsernameOrEmail

2018-09-29      Rick Richardson <rick.richardson@comcast.net>
	* geo-map: Change the default to -a leaflet
		This is because Google Maps recently puts:
		    "For development purposes only"
		on a map if you don't have an API Key.
	* geo-map: Add GMAP_KEY to the ~/.georc file in case you want to pay
		for google maps. Use MAPSRC=gmap in .georc.

2018-09-16      Dylan Richardson <dylanjackrichardson@gmail.com>
	* geo-map: Added Leaflet code, similar to -agmap.
		geo-map -a50 OR geo-map -aleaflet

2018-08-29      Rick Richardson <rick.richardson@comcast.net>
	* New program: radio-orphan-annie: Radio Orphan Annie’s Decoder Ring

2018-08-14      Rick Richardson <rick.richardson@comcast.net>
	* geo-common-gc:	GC.com changed protocol.
		Change:
		        URL="$GEOS/account/login"
		To:
			URL="$GEOS/account/signin"

2018-04-12      Rick Richardson <rick.richardson@comcast.net>
	* geo-common-gc: Add:
		-d 'ctl00$ContentBody$uxDownloadLoc=Download+Waypoints'
		to the "curl $URL #loc" for basic members.
	* geo-common-gc: Basic members: do it the real slow way!

2018-02-14      Rick Richardson <rick.richardson@comcast.net>
	* geo-html2gpx: Additional Waypoints fixed due to GC.com changes

2017-12-11      Rick Richardson <rick.richardson@comcast.net>
	* geo-common: GC.com changed protocol.
		get_hidden_value: sometimes they are two, thus the "tail -1"

2017-11-04      Rick Richardson <rick.richardson@comcast.net>
	* tap-code: Added -6. Matrix is 6x6, A..Z and 0..9

2017-10-25      Rick Richardson <rick.richardson@comcast.net>
	* addletters: Implement umlauts and -u

2017-05-14      tudado <creek@runbox.com>
	* geo-html2gpx: Update XML declaration

2017-04-28      tudado <creek@runbox.com>
	* geo-common-gc: Synchronize patterns used to determine is it archived
	    between gc_query() and getcids().  

2017-04-20      Rick Richardson <rick.richardson@comcast.net>
	* geo-id2gccode: convert integers up to i = 30

2017-04-12      Rick Richardson <rick.richardson@comcast.net>
	* Add FreeBSD

2017-04-04      Rick Richardson <rick.richardson@comcast.net>
	* geo-found: works for basic members.
		(it always works for premium members!)

2017-03-31      Rick Richardson <rick.richardson@comcast.net>
	* geo-demand: Change attribute AbandonedBuilding to abandonedbuilding

2017-03-21      Rick Richardson <rick.richardson@comcast.net>
	* geo-clock-angle: Compute the clock angle.
			   You can give the time or the angle.

2017-03-21      Rick Richardson <rick.richardson@comcast.net>
	* geo-compare-images: Compare two images

2017-03-08      Rick Richardson <rick.richardson@comcast.net>
	* geo-login changed.  geo-nearest/geo-newest works for premium members.
	* geo-login changed.  geo-nearest/geo-newest works for basic members.

2017-02-28      Rick Richardson <rick.richardson@comcast.net>
	* reverse-wherigo: Modify for six lengths of 6. (bflentje hack)

2017-02-15      Rick Richardson <rick.richardson@comcast.net>
	* geo-polygon: add -M Center of gravity (midpoint) with weights

2017-01-23      Rick Richardson <rick.richardson@comcast.net>
	* east and west confused when the location is near the meridian

2016-11-23      Rick Richardson <rick.richardson@comcast.net>
	* reverse-wherigo: Modify for lengths of 6 or 20

2016-11-19	Rick Richardson <rick.richardson@comcast.net>
	* adddigits: -m Multiplicative Persistance

2016-11-13      Rick Richardson <rick.richardson@comcast.net>
	* geo-common: degdec2mindec n0/e0 or w0/e0 put the right symbol.

2016-11-05      Rick Richardson <rick.richardson@comcast.net>
	* tap-code: Added -u and -r

2016-11-02      Rick Richardson <rick.richardson@comcast.net>
	* geo-ocr: Do -l pig with two encodings

2016-10-29      Rick Richardson <rick.richardson@comcast.net>
	* urwigo-decode: Add &lt; &gt; and &amp; to earwigo

2016-10-16      Rick Richardson <rick.richardson@comcast.net>
	* geo-common-gc: Bug for GC.com changes:
	    Next button s/disabled/Disabled/

2016-10-05      Rick Richardson <rick.richardson@comcast.net>
	* geo-pocket-query-newest: Bug for GC.com changes.

2016-08-18      Rick Richardson <rick.richardson@comcast.net>
	* geo-html2gpx: removed ctrl-h from the output

2016-07-11      Rick Richardson <rick.richardson@comcast.net>
	* geo-common-gc: logfiles not downloaded
	    Bug for GC.com changes: More https instead of http ...


geo-gid.sh
#!/bin/bash

#
#	HACK ALERT!  This was quickly hacked from an existing script,
#	and needs to be merged back into the main codebase.
#

#
#	Donated to the public domain by Rick Richardson <rickrich@gmail.com>
#
#	Use at your own risk.  Not suitable for any purpose.  Not legal tender.
#
#	$Id: geo-gid.sh,v 1.38 2018/07/08 16:12:12 rick Exp $
#

PROGNAME="$0"

usage() {
	cat <<EOF
NAME
    `basename $PROGNAME` - Fetch data about geocaches by gc.com GID

SYNOPSIS
    `basename $PROGNAME` [options] gid ...

DESCRIPTION
    Fetch data about geocaches by gc.com GID.  Only works with caches that
    are active (not archived).

    Requires:
        A subscriber ($30/yr) login at http://www.geocaching.com.  Visit a
        cache page and click the "Download to EasyGPS" link at least once so
        you can read and agree to the license terms.  Otherwise, you will not
        get any waypoint data.

	curl		http://curl.haxx.se/
	gpsbabel	http://gpsbabel.sourceforge.net/

OPTIONS
        -c              Remove cookie file when done
        -f              Do not report any found or unavailable caches
        -m              Do not report any members-only caches
        -F              Report caches found by the login 'username' as unfound
        -s              Output short names for the caches (gpsbabel option)
        -u username     Username for http://www.geocaching.com
        -p password     Password for http://www.geocaching.com
        -o format       Output format, -o? for possibilities [gpsdrive]
                        plus "gpsdrive.sql" for direct insertion into MySQL DB
                        plus "map[,geo-map-opts]" to display a geo-map.
        -O filename     Output file, if not stdout
        -S              Alias for -o gpsdrive.sql
        -d              For -S, just delete selected records
        -P              For -S, purge all records of type -t Geocache*
        -t type         For -ogpsdrive.sql, the waypoint type [Geocache]
        -H htmldir      Also fetch the printable HTML pages (slowly)
        -L logdir       Also fetch the plain text log entries (slowly)
                        For -H or -L, the limit is 1500 updated caches/day.
        -! "lpr -Plp"   Print HTML pages
        -D lvl          Debug level [0]
        -U              Retrieve latest version of this script

DEFAULTS
        Defaults can also be set with variables in file $HOME/.georc:

            PASSWORD=password;  USERNAME=username;   SOC=0|1;
            LAT=latitude;       LON=logitude;        GEOMYSTERY=/dev/null;
            NUM=num;            OUTFMT=format;       BABELFLAGS=-s;
            SQLUSER=gast;       SQLPASS=gast;        SQLDB=geoinfo;

EXAMPLES
    geo-gid GC4TAX4

SEE ALSO
    geo-newest, geo-found, geo-placed, geo-nearest,
    $WEBHOME
EOF

	exit 1
}

#include "geo-common"
#include "geo-common-gc"
#include "geo-common-gpsdrive"

gid2id() {
    gid="${1/#GC/}"
    ((val=0))
    ((pos=0))
    case "$gid" in
    ?????)
	while ((pos < 5)); do
	    digit=`expr index 0123456789ABCDEFGHJKMNPQRTVWXYZ "${gid:$pos:1}"`
	    ((val=val*31+$digit-1))
	    ((++pos))
	done
	;;
    *)
	if [ "$gid" "<" "G000" ]; then
	    echo "$((0x$gid))"
	    return;
	fi
	while ((pos < 4)); do
	    digit=`expr index 0123456789ABCDEFGHJKMNPQRTVWXYZ "${gid:$pos:1}"`
	    ((val=val*31+$digit-1))
	    ((++pos))
	done
	;;
    esac
    echo "$((val-=411120))"
}

# gid2id $1; exit

#
#	Query the gc website
#
gc_gid_query() {
    if [ $DEBUG -gt 0 ]; then
	TMP=/tmp/geo
    else
	TMP=/tmp/geo$$
    fi

    HTMLPAGE=$TMP.page
    CIDFILE=$TMP.cids
    LOCFILE=$TMP.loc
    XTRAFILE=$TMP.xtra
    CSVFILE=$TMP.csv
    MERGEFILE=$TMP.merge
    OUTWAY=$TMP.way
    STYLE=$TMP.newstyle

    CRUFT="$CRUFT $HTMLPAGE"
    CRUFT="$CRUFT $CIDFILE"
    CRUFT="$CRUFT $LOCFILE"
    CRUFT="$CRUFT $XTRAFILE"
    CRUFT="$CRUFT $CSVFILE"
    CRUFT="$CRUFT $MERGEFILE"
    CRUFT="$CRUFT $OUTWAY"
    CRUFT="$CRUFT $STYLE"
    if [ $NOCOOKIES = 1 ]; then
	CRUFT="$CRUFT $COOKIE_FILE"
    fi

    #
    #	Login to gc.com
    #
    gc_login "$USERNAME" "$PASSWORD"

    #
    #	We might combine one or more pages into a single XML, so cobble
    #	up a header with the ?xml and loc tags.
    #	
    cat <<-EOF > $LOCFILE
	<?xml version="1.0" encoding="UTF-8"?>
	<loc version="1.0" src="EasyGPS">
	EOF

    URL="$GEO/seek/nearest_cache.aspx"
    URL="$GEOS/seek/nearest.aspx"
    URL="$URL$SEARCH"
    dbgcmd curl $CURL_OPTS -L -s -b $COOKIE_FILE -A "$UA" \
	"$URL" > $HTMLPAGE
    rc=$?; if [ $rc != 0 ]; then
	error "curl: fetch $URL"
    fi

    ((qcnt=(GIDCNT+19)/20))
    ((q=0))
    ((i=0))
    > $XTRAFILE
    while ((q < qcnt)); do
	> $CIDFILE
	while ((i < GIDCNT)); do
	    gid=${GID[i]}
	    id=`gid2id "$gid"`
	    echo "-dCID=$id" >> $CIDFILE
	    echo "$gid	Geocache	0	0	0	0" >> $XTRAFILE
	    ((++i))
	    if ((i%20 == 0)); then
		break;
	    fi
	done
	((++q))

        #
        # Grab a few important values from the page
        #
	gc_getviewstate $HTMLPAGE

	dbgcmd curl $CURL_OPTS -s -b $COOKIE_FILE -A "$UA" \
		-d __EVENTTARGET="$__EVENTTARGET" \
		-d __EVENTARGUMENT="$__EVENTARGUMENT" \
		$viewstate \
		`cat $CIDFILE` \
		-d "Download=Download+Waypoints" \
		"$URL" \
	| sed -e 's/^<?xml [^>]*>//' \
		-e 's/>[gG]eocache</>Geocache</g' \
		-e 's/<loc [^>]*>//' \
		-e 's#</loc>##' \
		-e 's###g' \
		>> $LOCFILE
	rc=$?; if [ $rc != 0 ]; then
	    error "curl: fetch the waypoints"
	fi
	if grep -s -q "you are not logged in" $LOCFILE; then
	    error "you are not logged in on $start"
	fi
	if grep -s -q "has resulted in an error" $LOCFILE; then
	    error "searching error (3) on $start"
	fi
	if grep -s -q "Geocaching > Search for Geocaches" $LOCFILE; then
	    error "searching error (4) on $start"
	fi
	if grep -s -q "recaptcha_challenge_field" $LOCFILE; then
	    error "you are not subscription member!"
	fi

	#
	# Check to see if the user hasn't agreed to license terms
	#
	if grep -s -q "lblAgreementText" $LOCFILE; then
		easy_warning >&2
		remove_cruft
		exit
	fi
    done

    #
    # Finish off the .loc file
    #
    echo "</loc>" >> $LOCFILE

    #
    #	Convert the .loc data to .csv format and join it with
    #	the extra data scraped from the HTML page.  Filter out
    #	the data according to the -I and -X options.
    #
    # http://www.geocaching.com/seek/cache_details.aspx?wp=GCG2H4
    # http://www.geocaching.com/seek/cache_details.aspx?log=y&wp=GCG2H4
    # http://www.geocaching.com/seek/cache_details.aspx?ID=92117&log=y
    #
    # The joined .csv format looks like this:
    #	GCH636	Jidana 3 by rickrich	44.94520	-93.47540 \
    #	http://www.geocaching.com/seek/cache_details.aspx?log=y&wp=GCH636 \
    #	Geocache-ifound-regular	1070285077	1	0	1 \
    #	1067925600	1070285077	0
    #
    make_scrape_style > $STYLE

    dbgcmd gpsbabel -i geo$GEONUKE -f $LOCFILE \
	-o xcsv,style=$STYLE -F $CSVFILE
    if [ $? != 0 ]; then
	error "gpsbabel returned error code [1]"
    fi

    join -t '	' $CSVFILE $XTRAFILE \
    | egrep -- "$INCLUDE" | egrep -v -- "$EXCLUDE" \
    | sed -e 's/wp=/log=y\&&/' > $MERGEFILE

    #
    # Convert to the desired format
    #
    BABELFILT=
    if [ "$RADIUS" != "" ]; then
	BABELFILT="-x radius,distance=$RADIUS,lat=$LAT,lon=$LON"
    fi

    if [ $SQL = 1 ]; then
	    #
	    # add it via mysql
	    #
	    if [ "$OUTFILE" != "" ]; then
		>"$OUTFILE"
	    fi

	    if [ $PURGE = 1 ]; then
		gpsdrive_purge | gpsdrive_mysql
		PURGE=2
	    fi

	    dbgcmd gpsbabel $BABELFLAGS \
		-i xcsv,style=$STYLE -f $MERGEFILE \
		$BABELFILT -o "$OUTFMT" -F $OUTWAY
	    if [ $? != 0 ]; then
		error "gpsbabel returned error code [2]"
	    fi
	    gpsdrive_add <$OUTWAY $SQLTAG | gpsdrive_mysql
    elif [ $MAP = 1 ]; then
	    dbgcmd gpsbabel $BABELFLAGS \
		-i xcsv,style=$STYLE -f $MERGEFILE \
		$BABELFILT -o "$OUTFMT" -F $OUTWAY
	    if [ $? != 0 ]; then
		error "gpsbabel returned error code [3]"
	    fi
	    if [ "$OUTFILE" = "" ]; then
		dbgcmd geo-map -s0 $MAPOPTS -t$OUTWAY
	    else
		dbgcmd geo-map -s0 $MAPOPTS -t$OUTWAY -o"$OUTFILE"
	    fi
    else
	    #
	    # output to stdout or to a file
	    #
	    if [ "$OUTFILE" = "" ]; then
		OUTTMP="$TMP.way";  CRUFT="$CRUFT $OUTTMP"
		dbgcmd gpsbabel $BABELFLAGS \
		    -i xcsv,style=$STYLE -f $MERGEFILE \
		    $BABELFILT -o "$OUTFMT" -F $OUTTMP
		if [ $? != 0 ]; then
		    error "gpsbabel returned error code [4]"
		fi
		cat $OUTTMP
	    else
		dbgcmd gpsbabel $BABELFLAGS \
		    -i xcsv,style=$STYLE -f $MERGEFILE \
		    $BABELFILT -o "$OUTFMT" -F $OUTFILE
		if [ $? != 0 ]; then
		    error "gpsbabel returned error code [5]"
		fi
	    fi
    fi

    #
    #	Optionally, print the HTML pages
    #
    if [ "$CMDPIPE" != "" ]; then
        OIFS="$IFS"
        IFS="	"
        while read id desc lat lon url diff terr container strtype \
            vartime ifound soc iplaced placedt foundt ifoundt extra; do
            url="$url&decrypt=y"
            echo "Print: $url"
	    HTMLPAGE2=$TMP.html
	    CRUFT="$CRUFT $HTMLPAGE2"
	    debug 1 "curl $url" >&2
	    dbgcmd curl $CURL_OPTS -s -A "$UA" -b $COOKIE_FILE "$url" > $HTMLPAGE2
            htmldoc --quiet -t ps --nup 2 --fontsize 14 --webpage $HTMLPAGE2 \
                | psselect -q -p1-1 | eval $CMDPIPE
            # exit
        done < $MERGEFILE
        IFS="$OIFS"
    fi

    #
    #	Optionally, fetch printable HTML pages
    #
    if [ "$HTMLDIR" != ""  -o "$LOGDIR" != "" ]; then
	if [ "$HTMLDIR" != "" -a ! -d "$HTMLDIR" ]; then
	    mkdir "$HTMLDIR" || error "Couldn't mkdir $HTMLDIR"
	fi
	if [ "$LOGDIR" != "" -a ! -d "$LOGDIR" ]; then
	    mkdir "$LOGDIR" || error "Couldn't mkdir $LOGDIR"
	fi

	HTMLPAGE2=$TMP.html
	CRUFT="$CRUFT $HTMLPAGE2"

	OIFS="$IFS"
	IFS="	"
        while read id desc lat lon url diff terr container strtype \
            vartime ifound soc iplaced placedt foundt ifoundt extra; do
	    # Be kind to the server.  Do not remove this sleep
	    sleep $GEOSLEEP

            # Bug in gpsbabel 1.5.2 ...
            url=`echo "$url" | sed 's/.*http:/http:/g' `

	    gc_fetch_cache_page $HTMLPAGE $id $url ""
            if [ $? != 0 ]; then
                continue
            fi

	    if [ "$HTMLDIR" != "" ]; then
		cp $HTMLPAGE "$HTMLDIR/$id.html" ||
		    error "Couldn't copy $id cache page"
	    fi

	    if [ "$LOGDIR" != "" ]; then
		sed -e '1,/Cache find counts/d' -e 's/<STRONG><IMG SRC=/\
&/g' < $HTMLPAGE |
		egrep ">$LOGUSERNAME<|strong> \($LOGUSERNAME\) \(" > $HTMLPAGE2
		lynx -dump $HTMLPAGE2 > $LOGDIR/$id.log
	    fi
	done < $MERGEFILE
	IFS="$OIFS"

	if [ "$LOGDIR" != "" ]; then
	    # This is a hack, and might be innaccurate
	    printf "Finds:" >&2
	    MONTHS="  January |  February |  March |  April |  May |  June "
	    MONTHS="$MONTHS|  July |  August | September |  October "
	    MONTHS="$MONTHS|  November |  December "
	    egrep "$MONTHS" $LOGDIR/*.log |
	    egrep "icon_smile|icon_happy|icon_camera" | wc -l >&2
	fi

    fi
}

#
#       Set default options, can be overriden on command line or in rc file
#
UPDATE_URL=$WEBHOME/geo-nearest
UPDATE_FILE=geo-nearest.new

read_rc_file

#
#       Process the options
#

gc_getopts "$@"
shift $?

SEARCH="?origin_lat=$LAT&origin_long=$LON"

i=0
for gid in $*
do
    # Handle common typo of O for 0
    gid=$(echo "$gid" | tr 'O' '0')

    case "$gid" in
    GC*|gc*)
	GID[$i]=$gid
	((++i))
	;;
    *)	# GCID without the "GC"...
	GID[$i]="GC$gid"
	((++i))
	;;
    esac
done
GIDCNT=$i

gc_gid_query
